### ISSUE TITLE
[NodeBundle] Relating node to another node?

### ISSUE THREAD

**Issue by sfepoort** [Kunstmaan/KunstmaanBundlesCMS#2574]
Hi,

For a longer time I'm experiencing minor and major problems when relating nodes to nodes.
E.g: I try to link articlePages to articleCategoryPages (they have their own node/pages, it's not 'just an entity'). Or eventpages to contentpages, etc.

As far as i know, advice would be to relate 'your custom entity to the node'. Which is not an option in this setup. I've tried putting it on a page, which will end up in troubles when a new 'pageversion' gets created, by using drafts or other functions.

In short: How can I relate nodes to eachother, as in many-to-many (a setup with 'weight' would also really be helpfull for sorting). I assume i'm not the first one to run into this problem, but I can't find solutions in the demo-project, documentation, source-code or issue-list in here.

**Comment by acrobat** [comment]
@sfepoort If I understand your problem correctly this can be solved by adding a relation on the page object to the node entity. You will be able to select nodes on the admin interface by using the `Kunstmaan\NodeBundle\Form\NodeChoiceType`. The NodeChoiceType will display page titles in the correct language but will select nodes in the background. This way new pageversions won't be an issue as the main node object is linked.

Example

BlogPage entity
```php
class BlogPage 
{
    /**
     * @ORM\ManyToOne(targetEntity="Kunstmaan\NodeBundle\Entity\Node")
     */
    private $category;
}
```

BlogPageAdminType
```php
...
public function buildForm($builder, $options) {
    $builder->add('category', NodeChoiceType::class, [
        'page_class' => BlogCategoryPage::class,
    ]);
}
...
```

Does this solve your issue?

**Comment by sfepoort** [comment]
Hi @acrobat,

Thanks for your quick response! The ManyToOne does not offer a problem and works perfectly in our setup in the way you've described. :-)

I know this will also work with a many-to-many relation, by using the the field as
`@ORM\ManyToMany(targetEntity="Kunstmaan\NodeBundle\Entity\Node", orphanRemoval=false)` and providing the option multiple=>true to the NodeChoiceType for the form.

But in this case we like to add sorting in there, which is harder for me to grasp my thoughts around on what I need to build, and combine. Hope you will be as kind to direct me to a certain direction in this regard.

**Comment by sfepoort** [comment]
HI @acrobat ,

I've dived into it and we made a working proof-of-concept, so here to help others:
A working example of relating contentPageNodes to an ArticlePage, working with sorting by weight, including changes to all entities, new entity, forms and making sure versioining (drafts etc) are supported.

**The required new enttiy with the relation and weight**
[https://pastebin.com/11Ctfa78](https://pastebin.com/11Ctfa78)

**Changes to the articlePage which i want to relate contentNodes to**
[https://pastebin.com/r95Lg9SN](https://pastebin.com/r95Lg9SN)

**Changes to the articlePage form in the admin, to add a new collection for nodes**
[https://pastebin.com/DqyTUMbi](https://pastebin.com/DqyTUMbi)

**Individual formtype for the collection form-type, this is where we choose each individual node**
[https://pastebin.com/8NTbeZLx](https://pastebin.com/8NTbeZLx)

**Query to retrieve items**
Do not forget to write a proper query when retrieving items using these relationships.
**This is for a different setup then the above working proof of concept**, only need to change some naming in my opinion:
[https://pastebin.com/8aktSi1A](https://pastebin.com/8aktSi1A)

**Comment by acrobat** [comment]
@sfepoort thanks for the research! This would indeed be the way to go! I will tag this issue as documentation so we can port this to the docs in the future!
